### PR TITLE
Validation and testing

### PR DESCRIPTION
**Pull request by gabrielyea**
# Hello reviewer! :wave:  In this PR I have:
- Added validation to data models
- Added flash notifications for operations.
- Fixed all N+1 problems pointed out by bullet

## Notification on comment submitted:
flash-notif

### CODE REVIEW

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -1,5 +1,5 @@
-require 'rails_helper'
+# require 'rails_helper'
 
-RSpec.describe User, type: :model do
-  pending "add some examples to (or delete) #{__FILE__}"
-end
+# RSpec.describe User, type: :model do
+#   pending "add some examples to (or delete) #{__FILE__}"
+# end
```
Comment (ElijahTobs) [inline spec/models/user_spec.rb]: - [x] This file should contain validation tests for user model. Also, any custom model in the user.rb file should also be tested.

File: spec/models/post_spec.rb [hunk 2]
```diff
@@ -1,5 +1,49 @@
...
+    @post.comments_counter = nil
+    expect(@post).to_not be_valid
+  end
+
+  it 'should be invalid with negative values' do
+    @post.comments_counter = -10
+    expect(@post).to_not be_valid
+  end
+
+  it 'should be invalid with no likes counter' do
```
Comment (ElijahTobs) [inline spec/models/post_spec.rb]: - [x] The custom method(s) in the **`post.rb`** file is/are expected to have corresponding unit tests

File: spec/requests/likes_spec.rb [hunk 3]
```diff
@@ -1,7 +1,7 @@
...
-RSpec.describe 'Likes', type: :request do
-  describe 'GET /index' do
-    pending "add some examples (or delete) #{__FILE__}"
-  end
-end
+# RSpec.describe 'Likes', type: :request do
+#   describe 'GET /index' do
+#     pending "add some examples (or delete) #{__FILE__}"
+#   end
+# end
```
Comment (ElijahTobs) [inline spec/requests/likes_spec.rb]: - [x] The custom method(s) in the **`like.rb`** file is/are expected to have corresponding unit tests

File: spec/requests/comments_spec.rb [hunk 4]
```diff
@@ -1,7 +1,7 @@
...
-RSpec.describe 'Comments', type: :request do
-  describe 'GET /index' do
-    pending "add some examples (or delete) #{__FILE__}"
-  end
-end
+# RSpec.describe 'Comments', type: :request do
+#   describe 'GET /index' do
+#     pending "add some examples (or delete) #{__FILE__}"
+#   end
+# end
```
Comment (ElijahTobs) [inline spec/requests/comments_spec.rb]: - [x] The custom method(s) in the **`comment.rb`** file is/are expected to have corresponding unit tests
- [x] This file should also contain validation tests for comment model.

File: app/controllers/posts_controller.rb [hunk 5]
```diff
@@ -21,8 +21,10 @@ def create
     @post.likes_counter = 0
 
     if @post.save
+      flash[:notice] = 'Post saved!'
       redirect_to user_post_path(cu, @post)
     else
+      flash[:error] = @post.error.messages
       render :new
     end
   end
```
Comment (ElijahTobs) [inline app/controllers/posts_controller.rb]: - [x] You're expected to solve the N+1 problem when fetching all posts and their comments for a user. Hence, 

```ruby
@posts = @user.posts #should have been
@posts = @user.posts.includes(:comments)

**Review by ElijahTobs — CHANGES_REQUESTED** [review]
# Status: Changes Required ⚠️ 🚦
## Hi @gabrielyea  👋

You've done really well so far! 👍 

## Major Highlights 💯 

Your README is professional ✔️ 
You've followed the Gitflow rules. ✔️ 
You have good commit messages ✔️ 
You added validations for the input fields ✔️ 

_However, there are some changes that you are required to make before this project can be approved._

## Required Changes ♻️

- [ ] You should have `create` action for likes and have flash messages for successful and not successful path.


_See the inline comments below for more required changes_


_If anything doesn't seem clear to you, feel free to leave your question or comment in the PR thread below_

#### "Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise."

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
## Happy Coding :computer: 

_Reviewed by_  **Elijah O. Ayandokun**

File: spec/requests/comments_spec.rb (hunk 4 above)
Comment (gabrielyea) [reply spec/requests/comments_spec.rb]: Done!

File: spec/requests/likes_spec.rb (hunk 3 above)
Comment (gabrielyea) [reply spec/requests/likes_spec.rb]: Done!

**Review by descholar-ceo — APPROVED** [review]
Hi @gabrielyea , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have only 3 code reviews per project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**